### PR TITLE
00470 d fail invalid in regression

### LINKED ISSUE #470: FAIL_INVALID error in nightly regression
**Summary of the defect**
Observed the FAIL_INVALID errors again in nightly regression tests.

**How to reproduce (if possible)**
Check with CirclrCi's nightly regression summary URLs. For example, https://app.circleci.com/pipelines/github/hashgraph/hedera-services/2617/workflows/e4098ee0-0ec9-4511-a24c-b1cfe3ee9d1e/jobs/28551

**Service logs (if applicable)**
```

2020-08-22 09:54:06.824 INFO   125  ProviderRun - 40 minutes left in test - 27030 ops submitted so far (0 pending).
2020-08-22 09:54:06.824 INFO   132  ProviderRun - Precheck txn status counts :: {OK=11045}
2020-08-22 09:54:06.824 INFO   133  ProviderRun - Resolved txn status counts :: {INVALID_SIGNATURE=560, SUCCESS=10244, FILE_DELETED=237}
2020-08-22 09:54:06.824 INFO   134  ProviderRun - 
------------------------------

java.lang.AssertionError: Answer-only precheck was FAIL_INVALID, not one of [OK, INSUFFICIENT_TX_FEE, BUSY, INVALID_FILE_ID, PLATFORM_TRANSACTION_NOT_CREATED, FILE_DELETED]!
	at org.junit.Assert.fail(Assert.java:88)
	at com.hedera.services.bdd.spec.queries.HapiQueryOp.submitOp(HapiQueryOp.java:134)
	at com.hedera.services.bdd.spec.HapiSpecOperation.execFor(HapiSpecOperation.java:164)
	at com.hedera.services.bdd.spec.utilops.grouping.ParallelSpecOps.lambda$submitOp$1(ParallelSpecOps.java:51)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1807)
	at java.base/java.util.concurrent.CompletableFuture$AsyncRun.exec(CompletableFuture.java:1799)
	at java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)
	at java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)
	at java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)
	at java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)
	at java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:177)
2020-08-22 09:54:07.903 WARN   180  HapiSpecOperation - 'UmbrellaRedux' - ParallelSpecOps{numSubOps=10} failed!
java.lang.AssertionError: Problem(s) with sub-operation(s): HapiGetFileContents{sigs=1, payer=uniquePayerAccount, node=0.0.3, file=file1479-RandomFile} :: Answer-only precheck was FAIL_INVALID, not one of [OK, INSUFFICIENT_TX_FEE, BUSY, INVALID_FILE_ID, PLATFORM_TRANSACTION_NOT_CREATED, FILE_DELETED]!
	at org.junit.Assert.fail(Assert.java:88) ~[junit-4.12.jar:4.12]
	at com.hedera.services.bdd.spec.utilops.grouping.ParallelSpecOps.submitOp(ParallelSpecOps.java:64) ~[classes/:?]
	at com.hedera.services.bdd.spec.HapiSpecOperation.execFor(HapiSpecOperation.java:164) ~[classes/:?]
	at com.hedera.services.bdd.spec.utilops.CustomSpecAssert.allRunFor(CustomSpecAssert.java:38) ~[classes/:?]
	at com.hedera.services.bdd.spec.utilops.CustomSpecAssert.allRunFor(CustomSpecAssert.java:47) ~[classes/:?]
	at com.hedera.services.bdd.spec.utilops.ProviderRun.submitOp(ProviderRun.java:147) ~[classes/:?]
	at com.hedera.services.bdd.spec.HapiSpecOperation.execFor(HapiSpecOperation.java:164) ~[classes/:?]
	at com.hedera.services.bdd.spec.HapiApiSpec.exec(HapiApiSpec.java:216) ~[classes/:?]
	at com.hedera.services.bdd.spec.HapiApiSpec.run(HapiApiSpec.java:176) ~[classes/:?]
	at java.util.AbstractList$RandomAccessSpliterator.forEachRemaining(AbstractList.java:720) ~[?:?]
	at java.util.stream.ReferencePipeline$Head.forEach(ReferencePipeline.java:658) ~[?:?]
	at com.hedera.services.bdd.suites.HapiApiSuite.runSync(HapiApiSuite.java:233) ~[classes/:?]
	at com.hedera.services.bdd.suites.HapiApiSuite.runSuite(HapiApiSuite.java:125) ~[classes/:?]
	at com.hedera.services.bdd.suites.HapiApiSuite.runSuiteSync(HapiApiSuite.java:118) ~[classes/:?]
	at com.hedera.services.bdd.suites.SuiteRunner.lambda$runSuitesSync$14(SuiteRunner.java:352) ~[classes/:?]
	at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:176) ~[?:?]
	at java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:948) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) ~[?:?]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) ~[?:?]
	at com.hedera.services.bdd.suites.SuiteRunner.runSuitesSync(SuiteRunner.java:353) ~[classes/:?]
	at com.hedera.services.bdd.suites.SuiteRunner.lambda$runTargetCategories$10(SuiteRunner.java:326) ~[classes/:?]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) [?:?]
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1654) [?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) [?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) [?:?]
	at java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:913) [?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234) [?:?]
	at java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:578) [?:?]
	at com.hedera.services.bdd.suites.SuiteRunner.runTargetCategories(SuiteRunner.java:326) [classes/:?]
	at com.hedera.services.bdd.suites.SuiteRunner.runCategories(SuiteRunner.java:295) [classes/:?]
	at com.hedera.services.bdd.suites.SuiteRunner.main(SuiteRunner.java:235) [classes/:?]
	at org.codehaus.mojo.exec.ExecJavaMojo$1.run(ExecJavaMojo.java:254) [exec-maven-plugin-3.0.0.jar:?]
	at java.lang.Thread.run(Thread.java:835) [?:?]


```
**Environment:**
- OS: [e.g. Ubuntu 18.04]
- Java: [e.g. OpenJDK 11.0.4]
- Hedera Services Version: [e.g. 0.0.5]
- HAPI Version: [e.g. 0.0.5]

**Additional Context**
Add any other context about the problem here. Attach any logs here, if applicable.

### PR DESCRIPTION
**Pull request by ljianghedera**
**Related issue(s)**:
Closes #470 

**Summary of the change*

1. Added retry logic for file querying to avoid possible temporary failure querying binary object stores;
2. Also added log info to help debugging;

**External impacts**:
None.

**Applicable documentation**
- [ ] Javadoc
- [ ] HAPI protobuf comments
- [ ] README

### CODE REVIEW

File: test-clients/src/main/java/com/hedera/services/bdd/suites/crypto/CryptoCreateSuite.java [hunk 1]
```diff
@@ -119,7 +118,8 @@ public static HapiApiSpec invalidDurationGetsMeaningfulResponse() {
 	private HapiApiSpec vanillaCreateSucceeds() {
 		return defaultHapiSpec("VanillaCreateSucceeds")
 				.given(
-						cryptoCreate("testAccount").via("txn")
+						cryptoCreate("testAccount").via("txn"),
+						UtilVerbs.sleepFor(2000)
```
Comment (anighanta) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/crypto/CryptoCreateSuite.java:122]: This was passing without the sleep right ?

File: hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java [hunk 2]
```diff
@@ -58,6 +59,7 @@
 	private static final Logger log = LogManager.getLogger(StateView.class);
 
 	private static final byte[] EMPTY_BYTES = new byte[0];
+	private static final int RETRY_TIMES = 5;
```
Comment (Neeharika-Sompalli) [inline hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java:62]: Should the retries be 0 by default and should be dynamically configured by the test? So, that it doesn't affect any other places

File: test-clients/src/main/java/com/hedera/services/bdd/spec/HapiSpecOperation.java [hunk 3]
```diff
@@ -177,7 +177,7 @@ private AccountID randomNodeFrom(HapiApiSpec spec) {
 			}
 		} catch (Throwable t) {
 			if (!loggingOff) {
-				log.warn(spec.logPrefix() + this + " failed!", t);
+				log.warn(spec.logPrefix() + this.toString() + " failed!", t);
```
Comment (anighanta) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/HapiSpecOperation.java:180]: wouldn't this toString automatically called ?

File: test-clients/src/main/java/com/hedera/services/bdd/suites/crypto/CryptoCreateSuite.java (hunk 1 above)
Comment (ljianghedera) [reply test-clients/src/main/java/com/hedera/services/bdd/suites/crypto/CryptoCreateSuite.java:122]: Yeah, sometimes.

File: test-clients/src/main/java/com/hedera/services/bdd/spec/HapiSpecOperation.java (hunk 3 above)
Comment (ljianghedera) [reply test-clients/src/main/java/com/hedera/services/bdd/spec/HapiSpecOperation.java:180]: Yeah. It should. I added more log info and forget to clean this up. Fixed now.

File: hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java [hunk 4]
```diff
@@ -125,24 +127,38 @@ public StateView(
 	}
 
 	public Optional<FileGetInfoResponse.FileInfo> infoForFile(FileID id) {
-		try {
-			var attr = fileAttrs.get(id);
-			if (attr == null) {
+		int retries = RETRY_TIMES;
+		while (true) {
```
Comment (anighanta) [inline hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java:131]: I would recommend not using while true loops. I suggest to move this try block to a method that throws `BinaryObjectNotFoundException` and we catch it here and retry.

File: hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java (hunk 2 above)
Comment (ljianghedera) [reply hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java:62]: As matter of fact, I think all the queries to the binary object store should add the retry mechanism. 
Right now, only File query services run into this issue recently, but I have no doubt the same issue can happen to others given the right triggering condition.  For now,  let's just assume it only happens to File queries.

Right now, the re-try is only limited to file queries. Yeah, I will add this to properties to make it configurable.

File: hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java (hunk 4 above)
Comment (ljianghedera) [reply hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java]: Done with that. Still need a loop to retry, though not while(true) loop.

File: hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java [hunk 5]
```diff
@@ -125,25 +132,44 @@ public StateView(
...
-			var info = FileGetInfoResponse.FileInfo.newBuilder()
-					.setFileID(id)
-					.setDeleted(attr.isDeleted())
-					.setExpirationTime(Timestamp.newBuilder().setSeconds(attr.getExpirationTimeSeconds()))
-					.setSize(Optional.ofNullable(fileContents.get(id)).orElse(EMPTY_BYTES).length);
-			if (!attr.getWacl().isEmpty()) {
-				info.setKeys(mapJKey(attr.getWacl()).getKeyList());
+			retries--;
+			if (retries < 0) {
+				log.warn("Can't get info info for {} at this moment. Try again later", readableId(id));
```
Comment (anighanta) [inline hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java:152]: typo here

Comment (ljianghedera) [reply hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java:152]: Fixed.

**Pull request merged**